Implement secondary Mantle interface

ushitora-anqou opened the pull request.

toshipp commented on internal/controller/mantlebackup_controller.go:
```diff
@@ -327,32 +334,32 @@ func (r *MantleBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	}
 
-	if err := r.expire(ctx, &backup); err != nil {
+	if err := r.expire(ctx, backup); err != nil {
```
Should we do an expiration for all roles?

ushitora-anqou replied:
I think we should. I'll fix this point.

toshipp replied:
The expiration is a common task. Perhaps it is unacceptable for you, but I'd like you to suggest moving `expire` to `Reconcile` rather than each reconcile function.

toshipp replied:
We discussed this and concluded to keep it as is because we will refactor it not to use `WatchesRawSource`.

satoru-takeuchi reviewed (changes requested):
Although I'm in the middle of review, let me left some comments before completing the review to reduce the total review costs.

ushitora-anqou commented on internal/controller/mantlebackup_controller.go:
```diff
@@ -361,11 +368,57 @@ func (r *MantleBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request
+
+	if !isCreatedWhenMantleControllerWasSecondary(backup) {
+		logger.Info(
+			"skipping to reconcile the MantleBackup created by a different mantle-controller to prevent accidental data loss",
```
The word "remote" is a bit unclear here since the primary Ceph cluster is _remote_ from the secondary cluster's point of view. Maybe we should use "secondary" instead.

satoru-takeuchi replied:
I understood. Let's keep this message as is.

ushitora-anqou commented on internal/controller/mantlebackup_controller.go:
```diff
@@ -327,32 +334,32 @@ func (r *MantleBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request
 
 	if !backup.ObjectMeta.DeletionTimestamp.IsZero() {
-		return r.finalize(ctx, &backup, target, isErrTargetPVCNotFound(getSnapshotTargetErr))
+		return r.finalize(ctx, backup, target, isErrTargetPVCNotFound(getSnapshotTargetErr))
```
I agree with you. I'll fix this point.

ushitora-anqou replied:
And actually `finalizePrimary` is not necessary, because `ReoncileAsPrimary` calls `ReconcileAsStandalone`.

ushitora-anqou commented:
I updated this PR and added following three commits:

- add and use `finalizeStandalone` and `finalizeSecondary` instead of `finalize`.
- rename `ReconcileAs{Standalone,Primary,Secondary}` to `reconcileAs{...}`
- add expiration process to `reconcileAsSecondary`.

satoru-takeuchi commented on internal/controller/mantlebackup_controller.go:
```diff
@@ -610,6 +668,42 @@ func (r *MantleBackupReconciler) finalize(
+		return result, err
+	}
+
+	if !controllerutil.ContainsFinalizer(backup, MantleBackupFinalizerName) {
```
It should be before `secondaryCleanup()` to reduce the number of calls of `secondaryCleanup()`. Although it's not about secondary Mantle interface, L593 should be before L588. Could you fix both?

ushitora-anqou replied:
Fixed in  [7754592](https://github.com/cybozu-go/mantle/pull/53/commits/775459292a5ebe9edd23b10e5b0d704853cc2e52)

satoru-takeuchi commented on internal/controller/mantlebackup_controller.go:
```diff
@@ -765,6 +859,107 @@ func (r *MantleBackupReconciler) export(
+	backup *mantlev1.MantleBackup,
+	target *snapshotTarget,
+) (ctrl.Result, error) { //nolint:unparam
+	if result, err := r.checkExportDataAlreadyUploaded(ctx, backup); err != nil || !result.IsZero() {
```
```suggestion
	if result, err := r.isExportDataAlreadyUploaded(ctx, backup); err != nil || !result.IsZero() {
```

It's to clarify this function returns true if the data has already been uploaded.

ushitora-anqou replied:
Fixed in [601de02](https://github.com/cybozu-go/mantle/pull/53/commits/601de02dd7d68d2a66a6f8612dda43a00bd2705f)

satoru-takeuchi reviewed (changes requested):
I finished my review and left some minor suggestions. I'll merge after reflecting these suggestions and @toshipp approved.

ushitora-anqou commented on internal/controller/mantlebackup_controller.go:
```diff
@@ -574,7 +632,7 @@ func isCreatedWhenMantleControllerWasSecondary(backup *mantlev1.MantleBackup) bo
 }
 
-func (r *MantleBackupReconciler) finalize(
+func (r *MantleBackupReconciler) finalizeStandalone(
```
Fixed in [e197b5a](https://github.com/cybozu-go/mantle/pull/53/commits/e197b5ade565f5fac375fdc37866f31bbf68721d)

ushitora-anqou commented:
@satoru-takeuchi I've fixed your points, so could you review this again?

satoru-takeuchi reviewed (approved):
Please rebase after Toshi's approval.

ushitora-anqou commented:
@satoru-takeuchi I rebased and pushed my code. There should be no changes: https://github.com/cybozu-go/mantle/compare/601de02dd7d68d2a66a6f8612dda43a00bd2705f..024e131aea77e5fa7c266accbed41260606b6dca

ushitora-anqou commented:
Thank you!
